packages/sls-env/src/__tests__/dependencies.test.ts

```diff
@@ -174,5 +174,89 @@ describe('serverless environment', () => {
+        switch (event.message) {
+          case 'client1':
+            return (_m: string, _n: string) => `Using client1!`
+          case 'client2':
+            return (_m: string, _n: string) => `Using client2!`
+          default:
+            throw new Error(`Cannot resolve dependency based on event message: ${event.message}`)
+        }
+      }
+
+      const handler = environment<
+        Handler<MessageEvent, NameContext, string>,
+        any,
+        BuildMessageDependencies,
+        EventPayload
+      >()
+        .global({
+          buildMessage: setDependencyFactory<BuildMessage, EventPayload>(buildMessagesFactory())
+        })
+        .app(({ payload: { event, context }, dependencies: { buildMessage } }) =>
+          buildMessage(event.message, context.name)
+        )
+        .successHandler(({ result }) => result as string).start
+
+      const response1 = await handler({ message: 'client1' }, { name: 'hello' })
+      expect(response1).toBe('Using client1!')
+
+      const response2 = await handler({ message: 'client2' }, { name: 'hello' })
+      expect(response2).toBe('Using client2!')
+
+      await expect(handler({ message: 'client3' }, { name: 'world' })).rejects.toThrow(
+        new Error(`Cannot resolve dependency based on event message: client3`)
+      )
+    })
+
+    it('correctly resolves dependency factories dependent on context', async () => {
+      type BuildMessage = (message: string, name: string) => string
+      type BuildMessageDependencies = {
+        buildMessage: DependencyFactory<BuildMessage>
+      }
+
+      const buildMessagesFactory = () => ({ context }: EventPayload) => {
+        switch (context.name) {
+          case 'client1':
+            return (_m: string, _n: string) => `Using client1!`
+          case 'client2':
+            return (_m: string, _n: string) => `Using client2!`
+          default:
+            throw new Error(`Cannot resolve dependency based on event message: ${context.name}`)
+        }
+      }
+
+      const handler = environment<
+        Handler<MessageEvent, NameContext, string>,
+        any,
+        BuildMessageDependencies,
+        EventPayload
+      >()
+        .global({
+          buildMessage: setDependencyFactory<BuildMessage, EventPayload>(buildMessagesFactory())
+        })
+        .app(({ payload: { event, context }, dependencies: { buildMessage } }) =>
+          buildMessage(event.message, context.name)
+        )
+        .successHandler(({ result }) => result as string).start
+
+      const response1 = await handler({ message: 'hello' }, { name: 'client1' })
+      expect(response1).toBe('Using client1!')
+
+      const response2 = await handler({ message: 'hello' }, { name: 'client2' })
+      expect(response2).toBe('Using client2!')
+
+      await expect(handler({ message: 'hello' }, { name: 'client3' })).rejects.toThrow(
+        new Error(`Cannot resolve dependency based on event message: client3`)
+      )
+    })
```

Comment:
we are missing a test when it fails to resolve the dependency, also now sure if that should be responsibility of dependency factory function or rather `sls` 

if value returns undefined/null ideally `sls` would return error

Reply:
Isnt the last expect() checking exactly that?
The factory could not resolve the dependency with that event (`{ message: 'hello' }`) so it thew an Error.
The sls successfully caught that error and threw it again

I will add a check that the errorHandler was successfully called and move to separate test to be more explicit